test_rmw_implementation/test/test_subscription.cpp

```diff
@@ -377,3 +377,146 @@ TEST_F(CLASSNAME(TestSubscriptionUse, RMW_IMPLEMENTATION), count_mismatched_subs
+  EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
+  ret =
+    rmw_take_sequence(sub, count, &sequence_zero_count, &info_sequence, &taken, null_allocation);
+  EXPECT_EQ(RMW_RET_INVALID_ARGUMENT, ret) << rmw_get_error_string().str;
+  EXPECT_EQ(sequence_zero_count, original_zero_count);
+  EXPECT_EQ(info_sequence, original_info);
+  EXPECT_EQ(taken, 0u);
+  rmw_reset_error();
+  ret = rmw_message_sequence_fini(&sequence_zero_count);
+  EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
+  ret = rmw_message_sequence_fini(&original_zero_count);
+  EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
+
+  rmw_message_info_sequence_t info_sequence_zero_count =
+    rmw_get_zero_initialized_message_info_sequence();
+  rmw_message_info_sequence_t original_info_zero_count = info_sequence_zero_count;
+  ret = rmw_message_info_sequence_init(&info_sequence_zero_count, 0u, &allocator);
+  EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
+  ret = rmw_message_info_sequence_init(&original_info_zero_count, 0u, &allocator);
+  EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
+  ret =
+    rmw_take_sequence(sub, count, &sequence, &info_sequence_zero_count, &taken, null_allocation);
+  EXPECT_EQ(RMW_RET_INVALID_ARGUMENT, ret) << rmw_get_error_string().str;
+  EXPECT_EQ(sequence, original_sequence);
+  EXPECT_EQ(info_sequence_zero_count, original_info_zero_count);
+  EXPECT_EQ(taken, 0u);
+  rmw_reset_error();
+  ret = rmw_message_info_sequence_fini(&info_sequence_zero_count);
+  EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
+  ret = rmw_message_info_sequence_fini(&original_info_zero_count);
+  EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
+
+  const char * implementation_identifier = sub->implementation_identifier;
+  sub->implementation_identifier = "not-an-rmw-implementation-identifier";
+  ret = rmw_take_sequence(sub, count, &sequence, &info_sequence, &taken, null_allocation);
+  EXPECT_EQ(RMW_RET_INCORRECT_RMW_IMPLEMENTATION, ret) << rmw_get_error_string().str;
+  EXPECT_EQ(sequence, original_sequence);
+  EXPECT_EQ(info_sequence, original_info);
+  EXPECT_EQ(taken, 0u);
+  rmw_reset_error();
+  sub->implementation_identifier = implementation_identifier;
+
+  ret = rmw_message_sequence_fini(&sequence);
+  EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
+
+  ret = rmw_message_sequence_fini(&original_sequence);
+  EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
+
+  ret = rmw_message_info_sequence_fini(&info_sequence);
+  EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
+
+  ret = rmw_message_info_sequence_fini(&original_info);
+  EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
+}
+
+bool operator==(const rmw_message_sequence_t & m1, const rmw_message_sequence_t & m2)
+{
+  return m1.size == m2.size;
+}
+
+bool operator==(const rmw_message_info_sequence_t & m1, const rmw_message_info_sequence_t & m2)
+{
+  return m1.size == m2.size;
+}
```

Comment:
@Lobotuerk nit: consider leaving a note explaining the rationale behind the checks.

Reply:
Though I wonder if we shouldn't be simply checking for length equality in tests instead of adding operator overloads (since this isn't really comparing for equality).

Reply:
Updated on 4a8441f, went with just checking length, got rid of the operators==